tools/evict-shared-cache/evict-shared-cache.cpp

```diff
@@ -0,0 +1,199 @@
+#include "gopt/gopt.h"
+#include "util/poll.h"
+
+void print_help(const char* argv0) {
+  // clang-format off
+  std::cerr << std::endl
+     << "Usage: " << argv0 << " [OPTIONS]" << std::endl
+     << "  --cache  DIR     Evict from shared cache DIR"  << std::endl
+     << "  --policy POLICY  Evict using POLICY"           << std::endl
+     << "  --help   -h      Print this message and exit"  << std::endl
+     << "Commands (read from stdin):" << std::endl
+     << "  write JOB_ID     JOB_ID was written into the shared cache" << std::endl
+     << "  read JOB_ID      JOB_ID was read from the shared cache" << std::endl
+     << "Available Policies:" << std::endl
+     << "  nil              No op policy. Process commands but do nothing." << std::endl
+     << std::endl;
+  // clang-format on
+}
+
+std::unique_ptr<EvictionPolicy> make_policy(const char* argv0, const char* policy) {
+  if (strcmp(policy, "nil") == 0) {
+    return std::make_unique<NilEvictionPolicy>();
+  }
+
+  std::cerr << "Unknown policy: " << policy << std::endl;
+  print_help(argv0);
+  exit(EXIT_FAILURE);
+}
+
+enum class CommandParserState { Continue, StopSuccess, StopFail };
+
+struct CommandParser {
+  std::string command_buff = "";
+  Poll poll;
+
+  CommandParser() { poll.add(STDIN_FILENO); }
+
+  CommandParserState read_commands(std::vector<std::string>& commands) {
+    commands = {};
+
+    sigset_t saved;
+    struct timespec timeout;
+    timeout.tv_sec = 1;
+
+    // Sleep until timeout or a signal arrives
+    std::vector<int> ready_fds = poll.wait(&timeout, &saved);
+
+    // Nothing is ready within timeout, yield control
+    if (ready_fds.size() == 0) {
+      return CommandParserState::Continue;
+    }
+
+    while (true) {
+      uint8_t buffer[4096] = {};
+
+      ssize_t count = read(STDIN_FILENO, static_cast<void*>(buffer), 4096);
+
+      // Nothing new to process, yield control
+      if (count == 0) {
+        return CommandParserState::Continue;
+      }
+
+      // An error occured during read
+      if (count < 0) {
+        if (errno == EBADF) {
+          return CommandParserState::StopSuccess;
+        }
+
+        std::cerr << "Failed to read from stdin: " << strerror(errno) << std::endl;
+        return CommandParserState::StopFail;
+      }
+
+      // do the stuff
+      uint8_t* iter = buffer;
+      uint8_t* buffer_end = buffer + count;
+      while (iter < buffer_end) {
+        auto end = std::find(iter, buffer_end, 0);
+        command_buff.append(iter, end);
+        if (end != buffer_end) {
+          commands.emplace_back(std::move(command_buff));
+          command_buff = "";
+        }
+        iter = end + 1;
+      }
+
+      // last read consumed the buffer, yield control
+      if (count < 4096) {
+        return CommandParserState::Continue;
+      }
+
+      // yield if 100 commands have been buffered up without yielding
+      if (commands.size() > 100) {
+        return CommandParserState::Continue;
+      }
```

Comment:
Also here, no need to yield in this situation, just keep parsing. Since we can sleep indefinitely until a message comes along we don't need to yield ever. Just do as much work as we can in each place.

Reply:
This also enables bigger batch transactions which will be nice I think.